EAGLE-1335: New steps in checkGraph to check correctness of the graphConfigs

james-strauss-uwa opened the pull request:
I've added three things:

- Added graphConfigs to the list of things already checked (node, edge, field etc) to make sure all ids are unique
- Make sure the activeGraphConfigId is actually an id that belongs to a graphConfig in the graph
- Make sure that all fields, in all nodes, in all graphConfigs can actually be found in the LogicalGraph

Fixed bug that happened when you added a field to a graph config, then deleted the field from the LG node, or deleted the whole LG node. The graph config would then refer to a field/node that didn't exist, which would break the GraphConfigurationTable.

Now EAGLE shows a placeholder row for the "missing" field. See screenshot:

![Screenshot 2024-11-05 at 3 23 49 PM](https://github.com/user-attachments/assets/8602e13b-435b-49d8-b735-481f8407d0f7)

Also fixed a bug where the part of checkGraph that checked the LG itself was not called as part of Utils.checkGraph()

!!! **Please check** the way I modified ParameterTable.getNodeLockedState() to make the names in the GraphConfigurationAttributesTable readonly. Not sure if this is the best way.

## Summary by Sourcery

Enhance the checkGraph function to include validation for graphConfig IDs and ensure activeGraphConfigId is valid. Fix bugs related to non-existent field and node references in graph configurations, and improve the ParameterTable to display placeholders for missing fields.

New Features:
- Introduce checks to ensure all graphConfig IDs are unique and valid within the LogicalGraph.

Bug Fixes:
- Fix a bug where graph configurations could reference non-existent fields or nodes, causing errors in the GraphConfigurationTable.
- Resolve an issue where the checkGraph function did not validate the LogicalGraph itself.

Enhancements:
- Add placeholder rows for missing fields in the ParameterTable to handle cases where fields are deleted from the LogicalGraph.

james-strauss-uwa commented on src/LogicalGraph.ts:
```diff
@@ -901,13 +901,14 @@ export class LogicalGraph {
+        // clear old issues
+        graph.issues([]);
+
+        // check that all node, edge, field, and config ids are unique
```
I tried this, but it saved about 3 lines of code overall. Not worth the added complexity.

M-Wicenec commented on src/ParameterTable.ts:
```diff
@@ -199,6 +204,10 @@ export class ParameterTable {
     // TODO: move to Eagle.ts?
     //       doesn't seem to depend on any ParameterTable state, only Eagle state
     static getNodeLockedState = (field:Field) : boolean => {
+        if (Setting.find(Setting.BOTTOM_WINDOW_MODE).value() === Eagle.BottomWindowMode.GraphConfigAttributesTable){
```
i think this is fine. maybe add a comment that this is for the graph config table, the fields of which are always editable. 
We might have to revisit whether that is the case when we revamp ui-modes/permissions in the future

james-strauss-uwa replied:
I've added a comment to the code:

`// this handles a special case where EAGLE is displaying the "Graph Configuration Attributes Table"
  // all the fields shown in that table should be locked (readonly)`

This code is meant to make the "Attribute Name" field in the Graph Configuration Attributes Table" readonly. Since it doesn't really make sense to be able to rename a field when looking at a graph config.

M-Wicenec commented on src/LogicalGraph.ts:
```diff
@@ -945,10 +945,78 @@ export class LogicalGraph {
+                    }
+                }
+            }
+        }
```
i like the way this scenario is handled. The dev console is quite noisy about not being able to find a node though. perhaps we dont need to show those warnings?

M-Wicenec replied:
oh, i also just noticed that when the table is showing the missing node and field in the graph configuration attributes table, the tooltips cause an error when hovering on the node name.

james-strauss-uwa replied:
Yeah, true, there are a lot of warnings.

I'm using findNodeById**Quiet**() here, but there are a few other places that generated warnings. I've now changed the html data-bindings to use findNodeByIdQuiet() too.

And also removed two console.warn() that were explicitly called when getTableFields() could not find a node or field.

james-strauss-uwa replied:
The tooltip error was caused by passing the eagleTooltip binding undefined html.

I've updated the eagleTooltip code to avoid that error.

M-Wicenec reviewed (changes requested):
nice work! just three minor comments.

M-Wicenec commented:
I just noticed as well. two of the columns in the graph config attr table disappear when you delete the node and its showing the missing node field. they should probably still be visible so you can tell by the comment what it was, and are able to delete it yourself from there?

![image](https://github.com/user-attachments/assets/f690271d-0694-486c-bbe9-8b1e5f8d1340)

james-strauss-uwa commented:
> I just noticed as well. two of the columns in the graph config attr table disappear when you delete the node and its showing the missing node field. they should probably still be visible so you can tell by the comment what it was, and are able to delete it yourself from there?
> 
> ![image](https://private-user-images.githubusercontent.com/44516278/383415385-f690271d-0694-486c-bbe9-8b1e5f8d1340.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA4NzU1MDQsIm5iZiI6MTczMDg3NTIwNCwicGF0aCI6Ii80NDUxNjI3OC8zODM0MTUzODUtZjY5MDI3MWQtMDY5NC00ODZjLWJiZTktOGIxZTVmOGQxMzQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA2VDA2NDAwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM4ZmVlNzYxMGRmY2I3ZDE1N2JkZmFlM2IzMjJhYzg5ZmUyYzk0ZjZhNTcyNjA3ZGFmZTFiMTQ2MjA2ZWZlZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qUF1ngoRsjPUOkGUGgxP6gzVdqOFi3JRZmYryIv6ZKM)

Thanks, I've just fixed this error. The two columns are now visible and contain correct information where possible.

I've also fixed the "remove" button in the "remove" column, so that it won't error when removing the field, and will re-check the graph after removing the field.

james-strauss-uwa commented:
I think I've addressed all the issues you found, so it should be ready for another review.

M-Wicenec reviewed (approved):
yep, looks all good now
